lib/broadway_cloud_pub_sub/client.ex

```diff
@@ -0,0 +1,18 @@
+defmodule BroadwayCloudPubSub.Client do
+  @moduledoc """
+  A generic behaviour to implement Pub/Sub clients for `BroadwayCloudPubSub.Producer`.
+
+  This module defines callbacks to normalize options and receive messages
+  from a Cloud Pub/Sub topic. Modules that implement this behaviour should be passed
+  as the `:client` option from `BroadwayCloudPubSub.Producer`.
+  """
+
+  alias Broadway.Message
+
+  @type messages :: [Message.t()]
+  @type ack_ids :: [String.t()]
+
+  @callback init(opts :: any) :: {:ok, normalized_opts :: any} | {:error, message :: binary}
+  @callback acknowledge(ack_ids, opts :: any) :: no_return
+  @callback pull(demand :: pos_integer, opts :: any) :: messages
```

Comment:
I would like to propose to keep the previous behaviour (for now) for two reasons:

1. Align with SQS. In particular, I believe we avoided `pull` and used a more vague `receive_messages` to not impose any format

2. We are still debating how we are going to make Acks replaceable/composable. I believe in some cases users may want to bypass the original ack mechanism. So if we have an official API, then customizing it acks should be done through it, rather than at the client level. Separating them also has the benefit that if I want to change only receive or acks, i don't have to implement both.

One thing that we are actually super glad with doing when it comes to broadway is to just wait because we are sure more patterns will emerge with time. :)

Thoughts?

Reply:
I'd love to be a part of the debate! :)

This was an effort to SRP the client down to "talks to Google".

Pub/Sub offers no built-in DLX mechanism, so with the current connector you have to handle errors very defensively or you'll just end up getting the same messages over and over.

Reply:
> This was an effort to SRP the client down to "talks to Google".

Yes, I thought so. I think this is generally great, I am just proposing to hold on a bit on this front. :) We can track it as an issue and only go with the rename so far.

Regarding the error handling, that should hopefully be handled by broadway?

And we will make sure to have the discussions on the broadway repo (that's where they happen generally!).

Reply:
> Regarding the error handling, that should hopefully be handled by broadway?

Sorry, I meant _failure_ handling, if that makes any difference.  Right now failed messages get ignored in the `ack/3` callback, which just means they'll time out and get redelivered.

https://github.com/plataformatec/broadway_cloud_pub_sub/blob/f6edbd2a3236cb211585c54433fa3fa145ec2324/lib/broadway_cloud_pub_sub/google_api_client.ex#L64

I can open a separate issue to discuss.

Reply:
Ah yes! I think in SQS they go to another queue, which then you have to process. Do we have this option for Google Cloud PubSub? (/cc @msaraiva).

Reply:
Yup, SQS has no `nack` so usually you configure a dead-letter queue and let the server handle it. Otherwise, it will be redelivered.

Reply:
@josevalim There is nothing configurable from the Cloud Pub/Sub side.

Here's an example of a manual implementation: https://medium.com/google-cloud/dead-letter-queues-simple-implementation-strategy-for-cloud-pub-sub-80adf4a4a800

Reply:
Technically speaking, GCP/S has no `nack`, either.  Instead, you `modifyAckDeadline` to `0` and the message will be redelivered immediately. But that's the only other way to affect message delivery - you can `acknowledge` or you can `modifyAckDeadline` to give yourself more time to acknowledge. :)

Reply:
So we want configurable failure, which is exactly what we need for Kafka. So let's discuss about how to address both and then we apply it consistently. We will probably have some news on the Kafka front only after ElixirConf though. :)